pkg/vm/vm_core.go

```diff
@@ -1,15 +1,173 @@
 package vm
 
-import "github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+import (
+	"errors"
+	"fmt"
+
+	"github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+)
+
+type VirtualMachineError struct {
+	Msg string
+}
+
+func (e *VirtualMachineError) Error() string {
+	return fmt.Sprintf(e.Msg)
+}
 
 // VirtualMachine represents the Cairo VM.
 // Runs Cairo assembly and produces an execution trace.
 type VirtualMachine struct {
 	RunContext  RunContext
-	currentStep uint
+	CurrentStep uint
 	Segments    memory.MemorySegmentManager
 }
 
-func NewVirtualMachine() *VirtualMachine {
-	return &VirtualMachine{Segments: *memory.NewMemorySegmentManager()}
+func NewVirtualMachine() VirtualMachine {
+	return VirtualMachine{Segments: memory.NewMemorySegmentManager()}
+}
+
+type Operands struct {
+	Dst memory.MaybeRelocatable
+	Res *memory.MaybeRelocatable
+	Op0 memory.MaybeRelocatable
+	Op1 memory.MaybeRelocatable
+}
+
+type OperandsAddresses struct {
+	Dst_addr memory.Relocatable
+	Op0_addr memory.Relocatable
+	Op1_addr memory.Relocatable
+}
+
+// ------------------------
+//  Deduced Operands funcs
+// ------------------------
+
+type DeducedOperands struct {
+	Operands uint8
+}
```

Comment:
We don't need this optimization, this vm's aim is to be complete, but simple enough to teach others how to build their own

Reply:
if we are going to make the builtins are we going to need it?

Reply:
builtins dont use this

Reply:
Its an opmimization, we can instead insert the operands into memory in the function that deduces them like the older code does